tasks/tasks.go

```diff
@@ -308,35 +343,28 @@ func (task *taskType) Age() time.Duration {
 }
 
 // returns true if the task has completed (successfully or not), false otherwise
-func (task *taskType) Completed() bool {
+func (task taskType) Completed() bool {
 	return task.Status.Code == TransferStatusSucceeded ||
 		task.Status.Code == TransferStatusFailed
 }
 
 // requests that the task be canceled
-func (task *taskType) Cancel() {
+func (task *taskType) Cancel() error {
```

Comment:
what does the `*` mean in front of an argument?

Reply:
It's a pointer. Go uses pass-by-value semantics (like C), and uses pointers instead of references (also like C). In "methods" (or "functions with receivers", if you are into computing history), you'll see [pointers in the receiver](https://go.dev/tour/methods/4) when the method/function is allowed to change the state of the receiver. If you use a non-pointer receiver, the changes don't stick because they are applied to a copy of the receiver.

Java and C++ have very reference-heavy practices in real-world projects, even though they technically use pass-by-value (because references in those languages are passed by value, just like pointers). Python itself (like Fortran!) uses pass-by-reference semantics. All of these circumstances have conspired to produce a developer population that IMO doesn't really understand how pass-by-value semantics work. In Go (as in C), one is forced to understand these semantics, which is a drawback for those who don't like understanding things, but a downstream benefit to those who do. :-)

Reply:
Aha!
Yep, developers nowadays don't know they're born -- getting away with not knowing the difference between pass-by-value and pass-by-reference, using loosely-typed languages, IDEs that find all your errors for you, using ChatGPT or Copilot to generate code for them... They have no idea what it was like back in the olden days of computer science when every bit and byte counted and you had to write loops out by hand because `for` hadn't been invented yet. <shaking walking stick emphatically>

Reply:
It all amounts to needing an überexpert to understand what all their code does when it doesn't do the right thing anyway. It's all very lopsided.